### PR TITLE
[web] Add canUpdateAsMatch to PersistedPlatformView.

### PR DESCRIPTION
**Pull request by ditman**
## Description

Override `canUpdateAsMatch` on the PersistedPlatformView class. This allows the framework to discard potential match candidates when their viewId's differ. (The current implementation would allow for partial matches and was causing trouble (see linked bug below)).

Slightly simplify the `update` method of the Platform View, since we don't have to worry about the case where viewIds are equal.

Remove leftover `@protected` annotations from the engine.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/65529

## Tests

I added the following tests:

* web_ui/test/engine/surface/platform_view_test.dart (new file)

Ran `felt test` and all unit tests seem to pass. Didn't wait for chromedriver tests.

(However I'm not sure if the tests I wrote are OK wrt initializing the PersistedPlatformViews by hand)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/platform_view.dart [hunk 1]
```diff
@@ -75,18 +75,28 @@ class PersistedPlatformView extends PersistedLeafSurface {
     }
   }
 
+  // Platform Views can only be updated if their viewId matches.
+  @override
+  bool canUpdateAsMatch(PersistedSurface oldSurface) {
+    if (super.canUpdateAsMatch(oldSurface)) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/platform_view.dart:81]: I wonder if the `PersistedSurface.canUpdateAsMatch` should be made `@mustCallSuper`.

File: lib/web_ui/lib/src/engine/html/platform_view.dart [hunk 2]
```diff
@@ -75,18 +75,28 @@ class PersistedPlatformView extends PersistedLeafSurface {
...
+      if (oldSurface is PersistedPlatformView) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/platform_view.dart:82]: `super.canUpdateAsMatch` already checks the type, so here we should just cast (which I think will be optimized away by dart2js).

File: lib/web_ui/test/engine/surface/platform_view_test.dart [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+// Sends a platform message to create a Platform View with the given id and viewType.
+Future<void> _createPlatformView(int id, String viewType) {
+  final completer = Completer<void>();
+  window.sendPlatformMessage(
+    'flutter/platform_views',
+    codec.encodeMethodCall(MethodCall(
+      'create',
+      <String, dynamic>{'id': id, 'viewType': viewType,},
```
Comment (yjbanov) [inline lib/web_ui/test/engine/surface/platform_view_test.dart:73]: Remove the trailing comma (or reformat one-line per key/value pair)

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/html/platform_view.dart (hunk 1 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/html/platform_view.dart:81]: Good point, let me check how many other overrides of this we have and see what've been doing there.

File: lib/web_ui/lib/src/engine/html/platform_view.dart (hunk 2 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/html/platform_view.dart:82]: I thought `is` checks didn't have much cost anymore. I'll remove this and call `(oldSurface as PersistedPlatformView).viewId`...

File: lib/web_ui/test/engine/surface/platform_view_test.dart (hunk 3 above)
Comment (ditman) [reply lib/web_ui/test/engine/surface/platform_view_test.dart:73]: Ah, I'm going to run dartfmt -w in this file, see what happens :P

File: lib/web_ui/lib/src/engine/html/platform_view.dart (hunk 2 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/html/platform_view.dart:82]: Done

File: lib/web_ui/test/engine/surface/platform_view_test.dart (hunk 3 above)
Comment (ditman) [reply lib/web_ui/test/engine/surface/platform_view_test.dart:73]: Done

File: lib/web_ui/lib/src/engine/html/platform_view.dart (hunk 1 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/html/platform_view.dart:81]: My platform_view code seemed like the only that was overriding this method. I think calling super here is beneficial, so I've added @mustCallSuper to the base method.

**Pull request merged**